plugins/field-date/src/field_date.js

```diff
@@ -26,380 +28,378 @@ goog.require('goog.i18n.DateTimeSymbols');
 goog.require('goog.i18n.DateTimeSymbols_he');
 goog.require('goog.ui.DatePicker');
 
-
 /**
  * Class for a date input field.
- * @param {string=} value The initial value of the field. Should be in
- *    'YYYY-MM-DD' format. Defaults to the current date.
- * @param {Function=} validator A function that is called to validate
- *    changes to the field's value. Takes in a date string & returns a
- *    validated date string ('YYYY-MM-DD' format), or null to abort the change.
- * @param {?(boolean|string)=} textEdit Whether to enable text editor.
- * @extends {Blockly.Field}
- * @constructor
+ * @extends {Blockly.FieldTextInput}
  */
-Blockly.FieldDate = function(
-    value = undefined, validator = undefined, textEdit = false) {
+class FieldDate extends Blockly.FieldTextInput {
   /**
-   * The default value for this field (current date).
-   * @type {*}
-   * @protected
+   * Class for a date input field. Derived from the Closure library date
+   * picker.
+   * @param {string=} value The initial value of the field. Should be in
+   *    'YYYY-MM-DD' format. Defaults to the current date.
+   * @param {Function=} validator A function that is called to validate
+   *    changes to the field's value. Takes in a date string & returns a
+   *    validated date string ('YYYY-MM-DD' format), or null to abort the
+   *    change.
+   * @param {?(boolean|string)=} textEdit Whether to enable text editor.
    */
-  Blockly.FieldDate.prototype.DEFAULT_VALUE =
-      new goog.date.Date().toIsoString(true);
+  constructor(value = undefined, validator = undefined, textEdit = false) {
+    super(value, validator);
+
+    /**
+     * The default value for this field (current date).
+     * @type {*}
+     * @protected
+     */
+    this.DEFAULT_VALUE = (new goog.date.Date()).toIsoString(true);
```

Comment:
As far as I can tell, setting `DEFAULT_VALUE` here is useless:

- It appears only to be read by the `Field` constructor, which has already been invoked.
- The `Field` constructor never uses `this.DEFAULT_VALUE`, but instead uses `new.target` to read `FieldDate.prototype.DEFAULT_VALUE` (which is not set, so falls back to `FieldTextInput.prototype.DEFAULT_VALUE`, which is `''`).

I don't think there's any reason to try to set this on the instance at all.

The minimally-invasive fix is probably to change the `super` call:

```
    super(value ?? (new goog.date.Date()).toIsoString(true), validator);
````

This (like the original design) has the issue that multiple `FieldDate`s on the same page could get different initial values if the page is being loaded at midnight.  To avoid that the default value could be a module-local variable or (since this is still a `goog.provides` module) a static property on the `FieldDate` constructor.

Since we're already making breaking changes here, I can't help but think it might be better to fix the whole design by having `Field`s get their initial value from a static property on the constructor or by calling a static or instance method.  But that would be a much larger and more invasive change.

Reply:
Yeah, that placement of default value is useless, as the tests told me.

The problem here is that default value is used by the base class if validation fails. So using that change to the super call doesn't work, because it doesn't handle cases where the input date was invalid.

The tests all pass if I just move it to a static property on the class. 

But as you pointed out, that fails if the page is loaded at midnight (or, more commonly, if the page is left open for a long time) and I'm very leery of making quiet changes in how default dates work. I'll look at the alternatives you suggested.